spyder/api/plugins.py

```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+    # --- API: Signals -------------------------------------------------------
+    # ------------------------------------------------------------------------
 
-    def __init__(self, main=None):
-        super(BasePluginWidget, self).__init__(main)
+    # Use this signal to display a message in the status bar.
+    # str: The message you want to display
+    # int: Amount of time to display the message
+    sig_show_status_message = Signal(str, int)
 
-        # Dockwidget for the plugin, i.e. the pane that's going to be
-        # displayed in Spyder for this plugin.
-        # Note: This is created when you call the `add_dockwidget`
-        # method, which must be done in the `register_plugin` one.
-        self.dockwidget = None
+    # Use this signal to inform another plugin that configuration changed.
+    sig_option_changed = Signal(str, object)
 
-    def add_dockwidget(self):
-        """Add the plugin's QDockWidget to the main window."""
-        super(BasePluginWidget, self)._add_dockwidget()
+    def __init__(self, parent, configuration=None):
+        super().__init__(parent)
+        self._initialize(parent, configuration)
 
-    def tabify(self, core_plugin):
+    # FIXME: consider renaming this to setup_plugin, so we can achieve some
+    # language consistency. With this we can then do post_setup_plugin
+    # to refer to things to do after setup, or when visible.
+    # Using register and then post_visible_setup is not clear.
+    def register_plugin(self):
         """
-        Tabify plugin next to one of the core plugins.
+        Register plugin in Spyder's main window and connect it to other
+        plugins.
+        """
+        pass
 
-        Parameters
-        ----------
-        core_plugin: SpyderPluginWidget
-            Core Spyder plugin this one will be tabified next to.
 
-        Examples
-        --------
-        >>> self.tabify(self.main.variableexplorer)
-        >>> self.tabify(self.main.ipyconsole)
+class SpyderPluginWidget(QWidget, PluginBaseMixin):
+    # --- Private attributes -------------------------------------------------
+    # ------------------------------------------------------------------------
+    _ALLOWED_AREAS = Qt.AllDockWidgetAreas
+    _LOCATION = Qt.LeftDockWidgetArea
+    _FEATURES = QDockWidget.DockWidgetClosable | QDockWidget.DockWidgetMovable
 
-        Notes
-        -----
-        The names of variables associated with each of the core plugins
-        can be found in the `setup` method of `MainWindow`, present in
-        `spyder/app/mainwindow.py`.
-        """
-        super(BasePluginWidget, self)._tabify(core_plugin)
+    # --- API: Optional attributes -------------------------------------------
+    # ------------------------------------------------------------------------
 
-    def get_font(self, rich_text=False):
-        """
-        Return plain or rich text font used in Spyder.
+    # Path for images relative to the plugin path
+    IMG_PATH = 'images'
```

Comment:
SpyderPlugin could also require this (for example to use specific images as part of a new toolbar). So please move it to PluginBaseMixin.

Reply:
I do not really understand this as most plugins should follow a convention.

```
spyder_plugin
    locale/
    images/
    widgets/
    api.py
    plugin.py
    confpage.py
```

The only case I see some utility in this is on packages that are not a spyder_plugin alone but some pakcage that also registers a plugin, and on which case the images folder might not necessarily be inside the plugin folder. 

Having said this yes, I agree that most of the things for the two types of spyder plugins are common except what is strictly necessary for using a dock.